File: py/examples/otel/custom_otel_example.py
#!/usr/bin/env python3

import os
import time

# Set environment variables
os.environ.setdefault("BRAINTRUST_PARENT", "project_name:otel-examples")
os.environ.setdefault("BRAINTRUST_OTEL_ENABLE_LLM_FILTER", "false")

from braintrust.otel import Processor
from openai import OpenAI
from opentelemetry import trace
from opentelemetry.instrumentation.openai import OpenAIInstrumentor
from opentelemetry.sdk.trace import TracerProvider

# Set up the tracer provider
provider = TracerProvider()
trace.set_tracer_provider(provider)

# Instrument OpenAI to automatically trace calls
OpenAIInstrumentor().instrument()

# Define a custom filter function
def my_custom_filter(span):
    """Keep spans that start with 'custom_' in addition to LLM spans."""
    if span.name.startswith("custom_"):
        return True
    return None  # Defer to default LLM filtering logic


# Create a single processor with all the available options
processor = Processor(
    api_url="https://api.braintrust.dev",  # Base URL for Braintrust API
    custom_filter=my_custom_filter,  # Custom filter function
)

# Add the processor to the tracer provider
provider.add_span_processor(processor)

# Create a tracer and generate some spans
tracer = trace.get_tracer(__name__)

print("Creating spans to demonstrate filtering behavior...")

# Create spans to test the filtering behavior
with tracer.start_as_current_span("custom.otel.example") as main_span:
    main_span.set_attribute("request_id", "12345")
    main_span.set_attribute("user_id", "demo-user")

    # Add a simple OpenAI call - this will be automatically traced by OpenTelemetry
    client = OpenAI()
    response = client.chat.completions.create(
        model="gpt-3.5-turbo", messages=[{"role": "user", "content": "Hello, world!"}], max_tokens=10
    )

    main_span.set_attribute("openai_response", response.choices[0].message.content)

    # This span will be kept (LLM-related)
    with tracer.start_as_current_span("microservice-call") as llm_span:
        llm_span.set_attribute("gen_ai.model", "gpt-4")
        llm_span.set_attribute("gen_ai.tokens", 150)
        time.sleep(0.1)

    # This span will be filtered out (not LLM-related)
    with tracer.start_as_current_span("database_query"):
        time.sleep(0.05)

# Force flush to ensure spans are sent
trace.get_tracer_provider().force_flush(30)

File: js/examples/otel/custom_otel_example.ts
// custom_otel_example.ts
//
// This example shows how to use BraintrustSpanProcessor with a custom TracerProvider.
//
// New pattern benefits:
//   - Clean separation: OTEL code in @braintrust/otel package
//   - No async imports or try-catch blocks needed
//   - Clear error messages if dependencies missing
//   - Smaller bundle size for non-OTEL users
//
import { BasicTracerProvider } from "@opentelemetry/sdk-trace-base";
import { trace } from "@opentelemetry/api";
import { BraintrustSpanProcessor } from "@braintrust/otel";

const provider = new BasicTracerProvider({
  spanProcessors: [
    // Add Braintrust span processor with filtering enabled
    new BraintrustSpanProcessor({
      parent: "project_name:otel_examples",
      filterAISpans: true,
    }),
  ],
});

trace.setGlobalTracerProvider(provider); // sets the global tracer provider

console.log(
  "OpenTelemetry BasicTracerProvider started with BraintrustSpanProcessor",
);
console.log("BRAINTRUST_API_KEY set:", !!process.env.BRAINTRUST_API_KEY);

const tracer = trace.getTracer("custom-braintrust-service", "1.0.0");

async function makeRequest() {
  return tracer.startActiveSpan("custom.user_request", async (rootSpan) => {
    rootSpan.setAttributes({
      "user.request": "custom_example",
      "request.timestamp": new Date().toISOString(),
    });

    // This span will be kept by the filter (gen_ai prefix)
    tracer.startActiveSpan("gen_ai.chat.completions", async (aiSpan) => {
      aiSpan.setAttributes({
        "gen_ai.system": "openai",
        "gen_ai.operation.name": "chat.completions",
        "gen_ai.model": "gpt-4",
      });
      aiSpan.end();
    });

    // This span will be kept by the filter (braintrust prefix)
    tracer.startActiveSpan("braintrust.evaluation", async (braintrustSpan) => {
      braintrustSpan.setAttributes({
        "braintrust.dataset": "test-dataset",
        "braintrust.experiment": "test-experiment",
      });
      braintrustSpan.end();
    });

    // This span will be filtered out (no matching prefix)
    tracer.startActiveSpan("database.query", async (dbSpan) => {
      dbSpan.setAttributes({
        "db.system": "postgresql",
        "db.statement": "SELECT * FROM users",
      });
      dbSpan.end();
    });

    rootSpan.end();
  });
}

// Run the example
async function runExample() {
  await makeRequest();
  // Wait a moment for spans to be processed and sent
  await new Promise((resolve) => setTimeout(resolve, 2000));
  console.log(
    "\nSpans sent to Braintrust! Check your dashboard at https://braintrust.dev",
  );
  console.log(
    "Note: Only root spans and spans with gen_ai.*, braintrust.*, llm.*, ai.*, or traceloop.* prefixes were sent to Braintrust",
  );
  await provider.shutdown();
}

runExample();
